### DISCUSSION TITLE
Bulk save all child objects in BusinessListBase

### DISCUSSION THREAD

**Discussion by RocketLabsDev** [MarimerLLC/csla#2828] [Questions] [answered]
Hello,

Is there a way to bulk save all child objects in a list?
Do update on the objects that need update and insert on the objects that need insert.

Using dbconnection and dbcommand or EntityFramework.

**Comment by rockfordlhotka** [comment, marked as answer]
Yes, there are various ways to do this. I think some might be discussed in the Using CSLA 4 book.

The basic idea is this: the parent object (the list) has a default implementation for updating its children, and you can override that method. You basically need to do the same thing the default implementation does, but you can do it differently.

The default implementation loops through the list of deleted items and asks each child to delete itself, then they are removed from the deleted list.

It then loops through the list of active items and asks each child to update itself (if IsDirty is true), or insert itself (if IsNew is true).

Normally you put the code to insert/update/delete a child in the child class's child data portal operation methods, so each child object can do its task when asked by the parent.

_However_, you can put empty implementations of the insert/update/delete operations in the child class (as they still need to be called - just do nothing). In your override of the update children method in your list class, you do sort of the same thing:

1. Loop through the DeletedList and build up your query with all the delete operations - then clear DeletedList
2. Loop through the active items in the list and build up your query with the insert/update operations - making sure to call the child operation as normal so each child's metastate gets updated correctly (but doesn't actually persist, because your child operation methods are empty)
3. Run your consolidated query to batch run the delete/insert/update operations (the deletes _must_ occur first, or some inserts could fail!)
   1. In the case of success, you are done - yea!
   2. In the case of failure things get a little ugly, because now your in-memory object graph no longer matches what's in the database - so you may want to throw an exception and force the client to reload the entire object graph

The other bit of complexity to deal with, is if inserting a child results in a new db identity value for the child, you'll need to get those new identity values and figure out how to update the correct in-memory objects with their new property values.

**Reply by RocketLabsDev** [reply]
Ok, found something Csla 4 book about this, will try it out. Thnak you!

**Comment by michaelcsikos** [comment]
Hi @RocketLabsDev we have done this with a batch processor class to automagically generate SQL statements for a `DbContext`. We generally use `Guid` primary keys which makes this much simpler. If you need to retrieve any database-generated values you can collect the primary keys as you go and run a command after executing all of the batch deletes, inserts, and updates, and then loop through all the children to set the updated values. In the attached zip you'll see the `DbContextBatchProcessor`, `DbContextExtensions`, `EfTableDefinition` classes.

In your `DbContext` class add a lazy-loaded batch processor:

```cs
DbContextBatchProcessor m_batchProcessor = null;

public DbContextBatchProcessor BatchProcessor
{
    get
    {
        if (m_batchProcessor == null)
            m_batchProcessor = new DbContextBatchProcessor(this);

        return m_batchProcessor;
    }
}
```

In the root object you create a `DbContext`, start a transaction and call `Child_Update` passing in anything you need to collect any primary keys that will need to have new `Timestamp` values retrieved, etc. Each child appends its own insert, update or delete command, e.g. `db.BatchProcessor.AppendInsertCommand(data)` where data is an instance of an EF entity. If you need to do anything more complicated the child (or root) can append a stored procedure, e.g. `db.BatchProcessor.AppendStoredProcedureCommand("dbo.WidgetDelete", Id)`.

In the root call `db.BatchProcessor.SaveChanges()`, complete the transaction, and then run the command to retrieve the updated `Timestamp` values, etc.

For some scenarios the performance gain is massive, as you're only doing a handful of round-trips instead of thousands.

If anything is unclear, let me know.

[DbContextBatchProcessorAndExtensions.zip](https://github.com/MarimerLLC/csla/files/8241020/DbContextBatchProcessorAndExtensions.zip)

**Reply by RocketLabsDev** [reply]
Thank you for this information, will probably go with the dbconnection and dbcommand but this was very interesting and I thing others who are already using EF will benefit from this.
